File: backend/app/agents/retriever.py
from app.clients.courtlistener import CourtListenerClient
from app.clients.gemini import GeminiClient
from app.models.legal import Citation, LegalSource

QUOTE_PROMPT = """You are a legal research assistant. Given a legal query and a case snippet, extract the single most relevant quote from the snippet that directly supports or addresses the query.

Query: {query}

Case: {case_name}
Snippet: {snippet}

Respond in this exact format:
QUOTE: <the exact quote from the snippet>
EXPLANATION: <one sentence on why this quote is relevant to the query>

If the snippet has no relevant content, respond:
QUOTE: none
EXPLANATION: none"""


def run(query: str, court: str = "scotus", limit: int = 5) -> list[LegalSource]:
    cl_client = CourtListenerClient()
    gemini_client = GeminiClient()

    try:
        results = cl_client.search_opinions(query, court=court, limit=limit)
        sources = []

        for result in results:
            snippet = result.get("snippet", "").strip()
            case_name = result.get("caseName", "Unknown")
            citations = result.get("citation", [])
            date_filed = result.get("dateFiled", "")
            url = result.get("absolute_url", "")

            year = None
            if date_filed:
                try:
                    year = int(date_filed[:4])
                except ValueError:
                    pass

            citation = Citation(
                case_name=case_name,
                citation=citations[0] if citations else None,
                court=result.get("court_id"),
                year=year,
                url=f"https://www.courtlistener.com{url}" if url else None,
            )

            if not snippet:
                sources.append(LegalSource(
                    citation=citation,
                    relevant_quote="",
                    relevance_explanation="No snippet available.",
                ))
                continue

            prompt = QUOTE_PROMPT.format(
                query=query,
                case_name=case_name,
                snippet=snippet,
            )
            response = gemini_client.generate(prompt)

            quote = ""
            explanation = ""
            for line in response.splitlines():
                if line.startswith("QUOTE:"):
                    quote = line[len("QUOTE:"):].strip()
                elif line.startswith("EXPLANATION:"):
                    explanation = line[len("EXPLANATION:"):].strip()

            if quote == "none":
                quote = ""
                explanation = "No relevant content found in snippet."

            sources.append(LegalSource(
                citation=citation,
                relevant_quote=quote,
                relevance_explanation=explanation,
            ))

        return sources
    finally:
        cl_client.close()

File: frontend/lib/agents/retriever.ts
import { searchOpinions } from '@/lib/courtlistener';
import { LegalSource, Citation } from '@/types/legal';
import { generateText } from '@/lib/gemini';

const TOPIC_PROMPT = `You are a legal research assistant. Given a landmark case name, output a short search phrase (4–8 words) describing the core legal issue or constitutional question the case addresses. This phrase will be used to search a legal opinion database for related precedents.

Case: {case_name}

Respond with only the search phrase, nothing else.`;

const QUOTE_PROMPT = `You are a legal research assistant. Extract the single most legally significant quote from the case snippet below — a direct statement of law, constitutional principle, or judicial reasoning relevant to the topic.

Topic: {topic}

Case: {case_name}
Snippet: {snippet}

Respond in this exact format:
QUOTE: <exact verbatim quote from the snippet>
EXPLANATION: <one sentence on why this quote matters for the topic>

If the snippet is only procedural boilerplate or contains no substantive legal reasoning at all, respond:
QUOTE: none
EXPLANATION: none`;

const extractSnippet = (result: any): string =>
{
	const opinions: any[] = result.opinions ?? [];
	return opinions
		.map((o: any) => (o.snippet ?? '').trim())
		.filter((s: string) => s.length > 0)
		.join('\n\n');
};

export const run = async (
	query: string,
	options?: { court?: string; limit?: number; offset?: number; beforeYear?: number }
): Promise<LegalSource[]> =>
{
	const court = options?.court ?? 'scotus';
	const limit = options?.limit ?? 5;
	const offset = options?.offset ?? 0;
	const beforeYear = options?.beforeYear;
	const page = Math.floor(offset / limit) + 1;

	const topicPhrase = (await generateText(TOPIC_PROMPT.replace('{case_name}', query))).trim();

	const raw = await searchOpinions(topicPhrase, { court, limit: limit + 10, offset: (page - 1) * limit });
	const normalizedQuery = query.toLowerCase().replace(/[^a-z0-9]/g, '');

	const results = raw
		.filter(r =>
		{
			const name = (r.caseName ?? '').toLowerCase().replace(/[^a-z0-9]/g, '');
			if(name === normalizedQuery) return false;
			if(!extractSnippet(r)) return false;
			if(beforeYear && r.dateFiled)
			{
				const year = parseInt(r.dateFiled.slice(0, 4), 10);
				if(!isNaN(year) && year >= beforeYear) return false;
			}
			return true;
		})
		.slice(0, limit);

	const sources: LegalSource[] = [];

	for(const result of results)
	{
		const snippet = extractSnippet(result);
		const caseName: string = result.caseName ?? 'Unknown';
		const citations: string[] = result.citation ?? [];
		const dateFiled: string = result.dateFiled ?? '';
		const absoluteUrl: string = result.absolute_url ?? '';

		let year: number | null = null;
		if(dateFiled)
		{
			const parsed = parseInt(dateFiled.slice(0, 4), 10);
			if(!isNaN(parsed)) year = parsed;
		}

		const citation: Citation = {
			case_name: caseName,
			citation: citations[0] ?? null,
			court: result.court_id ?? null,
			year,
			url: absoluteUrl ? `https://www.courtlistener.com${absoluteUrl}` : null,
		};

		const prompt = QUOTE_PROMPT
			.replace('{topic}', topicPhrase)
			.replace('{case_name}', caseName)
			.replace('{snippet}', snippet);

		const response = await generateText(prompt);

		const quoteMatch = response.match(/QUOTE:\s*([\s\S]+?)(?=\nEXPLANATION:|$)/);
		const explanationMatch = response.match(/EXPLANATION:\s*([\s\S]+?)$/);

		let quote = quoteMatch?.[1]?.trim() ?? '';
		let explanation = explanationMatch?.[1]?.trim() ?? '';

		if(quote === 'none' || quote === '')
		{
			quote = '';
			explanation = 'No relevant content found in snippet.';
		}

		sources.push({ citation, relevant_quote: quote, relevance_explanation: explanation });
	}

	return sources;
};
